appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockVisibleExtension.java

```diff
@@ -0,0 +1,281 @@
+package com.google.appinventor.client.editor.simple.components;
+
+import com.google.appinventor.client.Ode;
+import com.google.appinventor.client.editor.simple.SimpleEditor;
+import com.google.appinventor.client.utils.Promise;
+import com.google.appinventor.client.utils.ShadowRoot;
+import com.google.appinventor.client.utils.jstypes.*;
+import com.google.appinventor.client.utils.jstypes.Worker.MessageEvent;
+import com.google.appinventor.client.widgets.properties.EditableProperty;
+import com.google.appinventor.shared.simple.ComponentDatabaseInterface.MockInfo;
+import com.google.gwt.core.client.JsonUtils;
+import com.google.gwt.dom.client.Document;
+import com.google.gwt.dom.client.Element;
+import com.google.gwt.json.client.JSONBoolean;
+import com.google.gwt.json.client.JSONNumber;
+import com.google.gwt.json.client.JSONObject;
+import com.google.gwt.json.client.JSONString;
+import com.google.gwt.json.client.JSONValue;
+import com.google.gwt.user.client.Window;
+import com.google.gwt.user.client.ui.FlowPanel;
+import com.google.gwt.user.client.ui.HTMLPanel;
+import com.google.gwt.user.client.ui.Image;
+import com.google.appinventor.components.common.PropertyTypeConstants;
+
+public class MockVisibleExtension extends MockVisibleComponent {
+  private final String MOCK_RENDER_REQUEST = "render-request";
+
+  private final long projectId;
+  private final String packageName;
+  private final MockInfo mockInfo;
+
+  private Worker worker;
+  private String workerUrl;
+
+  private final ShadowRoot shadowRoot;
+
+  public MockVisibleExtension(
+      SimpleEditor editor,
+      String typeName,
+      Image iconImage,
+      String packageName,
+      MockInfo mockInfo) {
+    super(editor, typeName, iconImage);
+    Ode.CLog("MockVisibleExtension.constructor");
+
+    this.projectId = editor.getProjectId();
+    this.packageName = packageName;
+    this.mockInfo = mockInfo;
+
+    final FlowPanel shadowHost = new FlowPanel();
+    shadowHost.setStylePrimaryName("ode-MockVisibleExtensionHost");
+    shadowRoot = applyAttachShadow(shadowHost.getElement());
+
+    initializeMock()
+        .then0(
+            () -> {
+              refreshForm();
+              return null;
+            });
+    initComponent(shadowHost);
+  }
+
+  private Promise<Object> initializeMock() {
+    Ode.CLog("MockVisibleExtension.initWorker");
+    final String assetsBasePath = "assets/external_comps/" + packageName + "/";
+
+    final Promise<String> fetchScript = fetchFileContent(assetsBasePath + mockInfo.getScript());
```

Comment:
What if the script is missing? How are we handling that case?

Reply:
We aren't handling that yet. Do you have any suggestions on how it should be handled?